cpp/src/hash/md5_hash.cu

```diff
@@ -27,23 +27,314 @@
+    uint32_t g;
+    switch (j / 16) {
+      case 0:
+        F = (B & C) | ((~B) & D);
+        g = j;
+        break;
+      case 1:
+        F = (D & B) | ((~D) & C);
+        g = (5 * j + 1) % 16;
+        break;
+      case 2:
+        F = B ^ C ^ D;
+        g = (3 * j + 5) % 16;
+        break;
+      case 3:
+        F = C ^ (B | (~D));
+        g = (7 * j) % 16;
+        break;
+    }
+
+    uint32_t buffer_element_as_int;
+    std::memcpy(&buffer_element_as_int, hash_state->buffer + g * 4, 4);
+    F = F + A + md5_hash_constants[j] + buffer_element_as_int;
+    A = D;
+    D = C;
+    C = B;
+    B = B + __funnelshift_l(F, F, md5_shift_constants[((j / 16) * 4) + (j % 4)]);
+  }
+
+  hash_state->hash_value[0] += A;
+  hash_state->hash_value[1] += B;
+  hash_state->hash_value[2] += C;
+  hash_state->hash_value[3] += D;
+
+  hash_state->buffer_length = 0;
+}
+
+/**
+ * @brief Core MD5 element processing function
+ */
+template <typename TKey>
+void CUDA_DEVICE_CALLABLE md5_process(TKey const& key, md5_intermediate_data* hash_state)
+{
+  uint32_t const len  = sizeof(TKey);
+  uint8_t const* data = reinterpret_cast<uint8_t const*>(&key);
+  hash_state->message_length += len;
+
+  // 64 bytes are processed in each hash step
+  constexpr int md5_chunk_size = 64;
+  if (hash_state->buffer_length + len < md5_chunk_size) {
+    std::memcpy(hash_state->buffer + hash_state->buffer_length, data, len);
+    hash_state->buffer_length += len;
+  } else {
+    uint32_t copylen = md5_chunk_size - hash_state->buffer_length;
+
+    std::memcpy(hash_state->buffer + hash_state->buffer_length, data, copylen);
+    md5_hash_step(hash_state);
+
+    while (len > md5_chunk_size + copylen) {
+      std::memcpy(hash_state->buffer, data + copylen, md5_chunk_size);
+      md5_hash_step(hash_state);
+      copylen += md5_chunk_size;
+    }
+
+    std::memcpy(hash_state->buffer, data + copylen, len - copylen);
+    hash_state->buffer_length = len - copylen;
+  }
+}
 
 // MD5 supported leaf data type check
 bool md5_type_check(data_type dt)
 {
   return !is_chrono(dt) && (is_fixed_width(dt) || (dt.id() == type_id::STRING));
 }
 
-}  // namespace
+struct MD5ListHasher {
+  template <typename T, std::enable_if_t<is_chrono<T>()>* = nullptr>
+  void __device__ operator()(column_device_view data_col,
+                             size_type offset_begin,
+                             size_type offset_end,
+                             md5_intermediate_data* hash_state) const
+  {
+    cudf_assert(false && "MD5 Unsupported chrono type column");
+  }
 
-namespace detail {
+  template <typename T, std::enable_if_t<!is_fixed_width<T>()>* = nullptr>
+  void __device__ operator()(column_device_view data_col,
+                             size_type offset_begin,
+                             size_type offset_end,
+                             md5_intermediate_data* hash_state) const
+  {
+    cudf_assert(false && "MD5 Unsupported non-fixed-width type column");
+  }
```

Comment:
Is there a reason not to just capture both groups of unsupported types in a single SFINAE overload? The extra info in the error message is pretty minimal and doesn't seem all that useful.

Reply:
I'm going to punt on this. I am planning a follow-up PR to make the MD5 code use the same base as the SHA code I'm working on in #9215. I combined these unsupported SFINAE overloads in that hash function base class, so it will be resolved in that PR.